### ISSUE TITLE
Fetch requests are never aborted + memory leak

### ISSUE THREAD

**Issue by comur** [rauliyohmc/react-native-offline#60]
Hi,

There is (in my opinion) a big problem by fetching an url to check connectivity as requests are never aborted so if you have an interval of 10s and your timeout is 3s, the code in checkInternetAccess will resolve false but the request is still maintained in stack and waiting for response as in node-fetch you cannot specify a real timeout.
So the problem is when there is a real timeout, fetch requests still waiting in stack got a reject asynchroniously and resolve => isConnected to false but actually you've the connection back.

Anyway it causes a non stable comportement in application, or sometimes you've got a hundred of request responses at once.

Another problem is that it causes a memory leak in app due to hundreds of requests loaded in memory waiting response.

Maybe you should change fetch with something else which is able to properly handle real timeouts ?

For now because of this problem I'm unable to use this in production with checkConnectionInterval parameter and I cannot use it without this parameter as on both platforms (Android and iOs) sometimes I don't have isConnected set back to true when I loose connection and I won it back...

One possible solution would be to abort waiting request when timeout occurs but it's not possible neither with node-fetch :/

**Comment by rauliyohmc** [comment]
Hi @comur, thanks for reporting this. You are totally right about that.

The XMLHttpRequest API is built in to React Native and supports cancellation. Another option would be to use [axios](https://github.com/axios/axios) that depends on it and may have a nicer API. I'd advocate for the former to not have extra dependencies.

I have a snippet of a _promisified_ version of XMLHttpRequest that I am using in one project. We could probably use that to make the `HEAD` requests instead of `fetch`.

Would you like to put together a PR if I provide you the snippet? I think it'd be simple and wouldn't require much adjustments, just wiring up everything :)

I am having time constraints right now. Otherwise, I'll try to address it, but probably wouldn't be able to do it till next year.

**Comment by comur** [comment]
Hi @rauliyohmc,
Send me everything I will check it as I need to use it ASAP if you can merge it quickly :)
Thanks

**Comment by rauliyohmc** [comment]
Here you go :)

```js
type Options = {
  method?: string,
  url: string,
  params:
    | string
    | {
        [name: string]: string,
      },
  headers: Object,
  timeout?: number,
};

/**
 * Utility that promisifies XMLHttpRequest in order to have a nice API that supports cancellation.
 * @param method
 * @param url
 * @param params -> This is the body payload for POST requests
 * @param headers
 * @param timeout -> Timeout for rejecting the promise and aborting the API request
 * @returns {Promise}
 */
export default function makeHttpRequest(
  { method = 'get', url, params, headers, timeout = 10000 }: Options = {},
) {
  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();

    const tOut = setTimeout(() => {
      xhr.abort();
      resolve('timeout');
    }, timeout);

    xhr.open(method, url);
    xhr.onload = function onLoad() {
      if (this.status >= 200 && this.status < 300) {
        clearTimeout(tOut);
        resolve(xhr.response);
      } else {
        clearTimeout(tOut);
        reject({
          status: this.status,
          statusText: xhr.statusText,
        });
      }
    };
    xhr.onerror = function onError() {
      clearTimeout(tOut);
      reject({
        status: this.status,
        statusText: xhr.statusText,
      });
    };
    if (headers) {
      Object.keys(headers).forEach(key => {
        xhr.setRequestHeader(key, headers[key]);
      });
    }
    let requestParams = params;
    // We'll need to stringify if we've been given an object
    // If we have a string, this is skipped.
    if (requestParams && typeof requestParams === 'object') {
      requestParams = Object.keys(requestParams)
        .map(
          (key: string) =>
            `${encodeURIComponent(key)}=${encodeURIComponent(
              // $FlowFixMe
              requestParams[key],
            )}`,
        )
        .join('&');
    }
    xhr.send(params);
  });
}
```

**Comment by rauliyohmc** [comment]
Fixed by #61
